tests/test_bilateral_precise.py

```diff
@@ -376,12 +377,29 @@ def test_cpu_precised(self, test_case_description, sigmas, input, expected):
         # Ensure result are as expected
         np.testing.assert_allclose(output, expected, atol=1e-5)
 
+    @parameterized.expand(TEST_CASES)
+    def test_cpu_precise_backwards(self, test_case_description, sigmas, input, expected):
+
+        # Params to determine the implementation to test
+        device = torch.device("cpu")
+        fast_approx = False
+
+        # Prepare input tensor
+        input_tensor = torch.from_numpy(np.array(input)).to(dtype=torch.float, device=device)
+        input_tensor.requires_grad = True
+
+        # Prepare args
+        args = (input_tensor, *sigmas, fast_approx)
+
+        # Run grad check
+        gradcheck(BilateralFilter.apply, args, raise_exception=False)
```

Comment:
I changed this to `raise_exception=True` and ran the first test case, it gives some issues:
```
Input #0 requires gradient and is not a double precision floating point or complex. This check will likely fail if all the inputs are not of double precision floating point or complex. 
E
======================================================================
ERROR: test_cpu_precise_backwards_0_1_dimension_1_channel_low_spatial_sigma_low_color_sigma (__main__.BilateralFilterTestCaseCpuPrecise)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/parameterized/parameterized.py", line 533, in standalone_func
    return func(*(a + p.args), **p.kwargs)
  File "/Users/Documents/MONAI/tests/test_bilateral_precise.py", line 76, in test_cpu_precise_backwards
    gradcheck(BilateralFilter.apply, args, raise_exception=True)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 468, in gradcheck
    checkIfNumericalAnalyticAreClose(a, n, j)
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 450, in checkIfNumericalAnalyticAreClose
    'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
  File "/usr/local/anaconda3/envs/py37/lib/python3.7/site-packages/torch/autograd/gradcheck.py", line 367, in fail_test
    raise RuntimeError(msg)
RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[ 9.8348e-01, -3.7858e-05, -5.4570e-06,  0.0000e+00,  0.0000e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [ 0.0000e+00,  5.7414e-01,  2.7407e-01,  7.7695e-02,  0.0000e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [ 0.0000e+00,  3.4821e-01,  4.5187e-01,  3.4821e-01,  0.0000e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [ 0.0000e+00,  7.7695e-02,  2.7407e-01,  5.7414e-01,  0.0000e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00, -5.4570e-06, -3.7858e-05,  1.0133e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
          7.4173e-01,  3.9520e-01,  0.0000e+00,  0.0000e+00,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
          2.5827e-01,  5.3274e-01,  0.0000e+00,  7.2094e-02,  0.0000e+00],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
         -5.1159e-06, -2.8706e-05,  9.8348e-01, -2.8706e-05, -5.1159e-06],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
          0.0000e+00,  7.2094e-02,  0.0000e+00,  5.3274e-01,  2.5827e-01],
        [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00,
          0.0000e+00,  0.0000e+00,  0.0000e+00,  3.9520e-01,  7.4173e-01]])
analytical:tensor([[1.0000e+00, 1.2041e-06, 2.1476e-07, 0.0000e+00, 0.0000e+00, 0.0000e+00,
         0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
        [1.5872e-06, 9.9999e-01, 1.2041e-06, 2.1476e-07, 0.0000e+00, 0.0000e+00,
         0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
        [2.1476e-07, 1.2041e-06, 9.9999e-01, 1.2041e-06, 2.1476e-07, 0.0000e+00,
         0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 2.1476e-07, 1.2041e-06, 9.9999e-01, 1.5872e-06, 0.0000e+00,
         0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 2.1476e-07, 1.2041e-06, 1.0000e+00, 0.0000e+00,
         0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.0000e+00,
         1.2041e-06, 2.1476e-07, 0.0000e+00, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 1.5872e-06,
         9.9999e-01, 1.2041e-06, 2.1476e-07, 0.0000e+00],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 2.1476e-07,
         1.2041e-06, 9.9999e-01, 1.2041e-06, 2.1476e-07],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
         2.1476e-07, 1.2041e-06, 9.9999e-01, 1.5872e-06],
        [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
         0.0000e+00, 2.1476e-07, 1.2041e-06, 1.0000e+00]])


----------------------------------------------------------------------
Ran 1 test in 0.006s

FAILED (errors=1)

```

Reply:
So the backpropergation is very rough, I'm currently working on reimplementing it along with the backprop of the PHLFilter. In my experiments im still able to successfully train a network containing this iteration of the bilateral filter. It depends on wether we want to include this rough implementation or throw a non implemented exception in the backwards pass preventing people from using it within a network for now.